featurebyte/migration/service/feature.py

```diff
@@ -0,0 +1,137 @@
+        derive_primary_entity_helper: DerivePrimaryEntityHelper,
+    ):
+        super().__init__(persistent)
+        self.feature_service = feature_service
+        self.entity_service = entity_service
+        self.relationship_info_service = relationship_info_service
+        self.entity_relationship_extractor_service = entity_relationship_extractor_service
+        self.derive_primary_entity_helper = derive_primary_entity_helper
+
+    @property
+    def delegate_service(self) -> BaseDocumentServiceT:
+        return self.feature_service  # type: ignore[return-value]
+
+    async def batch_preprocess_document(self, documents: List[Document]) -> List[Document]:
+        """
+        Preprocess the documents before migration
+
+        Parameters
+        ----------
+        documents: List[Document]
+            List of documents to be migrated
+
+        Returns
+        -------
+        List[Document]
+        """
+        all_entity_ids = set()
+        for document in documents:
+            all_entity_ids.update(document["entity_ids"])
+
+        # get all entity first to reduce the number of queries
+        entity_id_to_entity: Dict[ObjectId, EntityModel] = {
+            entity.id: entity
+            async for entity in self.entity_service.list_documents_iterator(
+                query_filter={"_id": {"$in": list(all_entity_ids)}}
+            )
+        }
+        extractor = self.entity_relationship_extractor_service
+        for document in documents:
+            primary_entity_ids = await self.derive_primary_entity_helper.derive_primary_entity_ids(
+                entity_ids=document["entity_ids"], entity_id_to_entity=entity_id_to_entity
+            )
+            relationships_info = await extractor.extract_relationship_from_primary_entity(
+                entity_ids=document["entity_ids"],
+                primary_entity_ids=primary_entity_ids,
+            )
+            document["primary_entity_ids"] = sorted(primary_entity_ids)
+            document["relationships_info"] = [
+                relationship.dict(by_alias=True) for relationship in relationships_info
+            ]
+        return documents
+
+    def migrate_document_record(self, record: Dict[str, Any]) -> Dict[str, Any]:
+        try:
+            output = super().migrate_document_record(record)
+            return output
+        except Exception as exc:  # pylint: disable=broad-except
+            # for some bad records, we just log the error and return the original record
+            # so that the migration can continue
+            logger.info(
+                "Record (id: %s, name: %s) encountered error: %s",
+                record["_id"],
+                record["name"],
+                str(exc),
+            )
+            return record
+
+    @migrate(version=4, description="Add relationships_info to feature record")
+    async def add_relationships_info(self) -> None:
+        """Change field name from event_data_ids to tabular_data_ids"""
```

Comment:
do we need to update this docstring/

Reply:
Good catch, will update the docstring here.

Reply:
Will update the docstring in the coming PR.